File: shared/StreamingEndpoint.swift
import Foundation

// Pure-logic endpoint decisions. Audio storage and pre-roll buffering remain
// owned by StreamingAudioRecorder.

enum StreamingEndpointEvidence: String {
    case strong
    case continuing
    case ambiguous
    case silence
}

enum StreamingEndpointState: String {
    case idle
    case onsetPending
    case speechActive
    case endPending
}

enum StreamingEndpointFinalizeKind: String {
    case endpoint
    case stop
}

enum StreamingEndpointDecision {
    case none
    case startUtterance(withPreRollSamples: Int)
    case continueUtterance
    case finalizeUtterance(kind: StreamingEndpointFinalizeKind)
}

struct StreamingEndpointConfiguration {
    let onsetSamples: Int
    let endEvidenceSamples: Int
    let endpointSilenceSamples: Int
    let resumeSamples: Int
    let ambiguousRescueSamples: Int
    let preRollSamples: Int

    init(
        onsetSamples: Int = 1_120,
        endEvidenceSamples: Int = 1_600,
        endpointSilenceSamples: Int = 11_200,
        resumeSamples: Int = 1_920,
        ambiguousRescueSamples: Int = 5_120,
        preRollSamples: Int = 4_000
    ) {
        self.onsetSamples = onsetSamples
        self.endEvidenceSamples = endEvidenceSamples
        self.endpointSilenceSamples = endpointSilenceSamples
        self.resumeSamples = resumeSamples
        self.ambiguousRescueSamples = ambiguousRescueSamples
        self.preRollSamples = preRollSamples
    }
}

final class StreamingEndpoint {
    let configuration: StreamingEndpointConfiguration
    private(set) var state: StreamingEndpointState = .idle
    private(set) var utteranceDurationSamples = 0
    private(set) var accumulatedSilenceSamples = 0
    private(set) var resumeEvidenceSamples = 0
    private(set) var ambiguousEvidenceSamples = 0
    private(set) var onsetEvidenceSamples = 0

    init(configuration: StreamingEndpointConfiguration = StreamingEndpointConfiguration()) {
        self.configuration = configuration
    }

    func process(
        evidence: StreamingEndpointEvidence,
        durationSamples: Int
    ) -> StreamingEndpointDecision {
        let samples = max(0, durationSamples)
        switch state {
        case .idle:
            return processIdle(evidence: evidence, durationSamples: samples)
        case .onsetPending:
            return processOnset(evidence: evidence, durationSamples: samples)
        case .speechActive:
            return processActive(evidence: evidence, durationSamples: samples)
        case .endPending:
            return processEndPending(evidence: evidence, durationSamples: samples)
        }
    }

    func forceFinalize(kind: StreamingEndpointFinalizeKind) -> StreamingEndpointDecision {
        guard state != .idle else { return .none }
        guard state != .onsetPending else {
            reset()
            return .none
        }
        reset()
        return .finalizeUtterance(kind: kind)
    }

    func reset() {
        state = .idle
        utteranceDurationSamples = 0
        accumulatedSilenceSamples = 0
        resumeEvidenceSamples = 0
        ambiguousEvidenceSamples = 0
        onsetEvidenceSamples = 0
    }

    private func processIdle(
        evidence: StreamingEndpointEvidence,
        durationSamples: Int
    ) -> StreamingEndpointDecision {
        guard evidence == .strong else { return .none }
        state = .onsetPending
        onsetEvidenceSamples = durationSamples
        if onsetEvidenceSamples >= configuration.onsetSamples {
            return beginUtterance()
        }
        return .none
    }

    private func processOnset(
        evidence: StreamingEndpointEvidence,
        durationSamples: Int
    ) -> StreamingEndpointDecision {
        guard evidence == .strong else {
            reset()
            return .none
        }
        onsetEvidenceSamples += durationSamples
        if onsetEvidenceSamples < configuration.onsetSamples {
            return .none
        }
        return beginUtterance()
    }

    private func beginUtterance() -> StreamingEndpointDecision {
        state = .speechActive
        utteranceDurationSamples = configuration.preRollSamples + onsetEvidenceSamples
        accumulatedSilenceSamples = 0
        resumeEvidenceSamples = 0
        ambiguousEvidenceSamples = 0
        return .startUtterance(withPreRollSamples: configuration.preRollSamples)
    }

    private func processActive(
        evidence: StreamingEndpointEvidence,
        durationSamples: Int
    ) -> StreamingEndpointDecision {
        utteranceDurationSamples += durationSamples
        switch evidence {
        case .strong, .continuing:
            accumulatedSilenceSamples = 0
            return .continueUtterance
        case .ambiguous:
            return .continueUtterance
        case .silence:
            accumulatedSilenceSamples += durationSamples
            guard accumulatedSilenceSamples >= configuration.endEvidenceSamples else {
                return .continueUtterance
            }
            state = .endPending
            return .continueUtterance
        }
    }

    private func processEndPending(
        evidence: StreamingEndpointEvidence,
        durationSamples: Int
    ) -> StreamingEndpointDecision {
        utteranceDurationSamples += durationSamples
        switch evidence {
        case .ambiguous:
            ambiguousEvidenceSamples += durationSamples
            guard ambiguousEvidenceSamples >= configuration.ambiguousRescueSamples else {
                return .continueUtterance
            }
            state = .speechActive
            accumulatedSilenceSamples = 0
            resumeEvidenceSamples = 0
            ambiguousEvidenceSamples = 0
            return .continueUtterance
        case .silence:
            ambiguousEvidenceSamples = 0
            accumulatedSilenceSamples += durationSamples
            resumeEvidenceSamples = 0
            guard accumulatedSilenceSamples >= configuration.endpointSilenceSamples else {
                return .continueUtterance
            }
            reset()
            return .finalizeUtterance(kind: .endpoint)
        case .strong, .continuing:
            // Resume evidence pauses the silence timer but cannot cancel the
            // pending end until the resume threshold is reached.
            ambiguousEvidenceSamples = 0
            resumeEvidenceSamples += durationSamples
            guard resumeEvidenceSamples >= configuration.resumeSamples else {
                return .continueUtterance
            }
            state = .speechActive
            accumulatedSilenceSamples = 0
            resumeEvidenceSamples = 0
            return .continueUtterance
        }
    }
}

File: scripts/prediction-sim/lib/streaming-endpoint.mjs
// Pure-logic JS mirror of shared/StreamingEndpoint.swift. This module owns
// timing decisions only; the recorder owns audio buffers and pre-roll storage.

export const STREAM_ENDPOINT_SILENCE_MS = 700;

export const STREAM_ENDPOINT_ONSET_SAMPLES = 1120;
export const STREAM_ENDPOINT_END_EVIDENCE_SAMPLES = 1600;
export const STREAM_ENDPOINT_RESUME_SAMPLES = 1920;
export const STREAM_ENDPOINT_AMBIGUOUS_RESCUE_SAMPLES = 5120;
export const STREAM_ENDPOINT_PREROLL_SAMPLES = 8000;

export const StreamingEndpointEvidence = Object.freeze({
  strong: 'strong',
  continuing: 'continuing',
  ambiguous: 'ambiguous',
  silence: 'silence',
});

export const StreamingEndpointState = Object.freeze({
  idle: 'idle',
  onsetPending: 'onsetPending',
  speechActive: 'speechActive',
  endPending: 'endPending',
});

export const StreamingEndpointFinalizeKind = Object.freeze({
  endpoint: 'endpoint',
  stop: 'stop',
});

export function makeStreamingEndpointConfig(partial = {}) {
  const config = {
    onsetSamples: STREAM_ENDPOINT_ONSET_SAMPLES,
    endEvidenceSamples: STREAM_ENDPOINT_END_EVIDENCE_SAMPLES,
    endpointSilenceSamples: STREAM_ENDPOINT_SILENCE_MS * 16,
    resumeSamples: STREAM_ENDPOINT_RESUME_SAMPLES,
    ambiguousRescueSamples: STREAM_ENDPOINT_AMBIGUOUS_RESCUE_SAMPLES,
    preRollSamples: STREAM_ENDPOINT_PREROLL_SAMPLES,
    ...(partial ?? {}),
  };
  if (partial?.endpointSilenceMs !== undefined) {
    config.endpointSilenceSamples = partial.endpointSilenceMs * 16;
  }
  return config;
}

export class StreamingEndpoint {
  constructor(config = {}) {
    this.configuration = makeStreamingEndpointConfig(config);
    this.state = StreamingEndpointState.idle;
    this.utteranceDurationSamples = 0;
    this.accumulatedSilenceSamples = 0;
    this.resumeEvidenceSamples = 0;
    this.ambiguousEvidenceSamples = 0;
    this.onsetEvidenceSamples = 0;
  }

  process(evidence, durationSamples) {
    const samples = Math.max(0, durationSamples);
    switch (this.state) {
      case StreamingEndpointState.idle:
        return this.processIdle(evidence, samples);
      case StreamingEndpointState.onsetPending:
        return this.processOnset(evidence, samples);
      case StreamingEndpointState.speechActive:
        return this.processActive(evidence, samples);
      case StreamingEndpointState.endPending:
        return this.processEndPending(evidence, samples);
      default:
        return this.none();
    }
  }

  forceFinalize(kind) {
    if (this.state === StreamingEndpointState.idle) return this.none();
    if (this.state === StreamingEndpointState.onsetPending) {
      this.reset();
      return this.none();
    }
    this.reset();
    return this.finalize(kind);
  }

  reset() {
    this.state = StreamingEndpointState.idle;
    this.utteranceDurationSamples = 0;
    this.accumulatedSilenceSamples = 0;
    this.resumeEvidenceSamples = 0;
    this.ambiguousEvidenceSamples = 0;
    this.onsetEvidenceSamples = 0;
  }

  processIdle(evidence, durationSamples) {
    if (evidence !== StreamingEndpointEvidence.strong) return this.none();
    this.state = StreamingEndpointState.onsetPending;
    this.onsetEvidenceSamples = durationSamples;
    if (this.onsetEvidenceSamples >= this.configuration.onsetSamples) {
      return this.beginUtterance();
    }
    return this.none();
  }

  processOnset(evidence, durationSamples) {
    if (evidence !== StreamingEndpointEvidence.strong) {
      this.reset();
      return this.none();
    }
    this.onsetEvidenceSamples += durationSamples;
    if (this.onsetEvidenceSamples < this.configuration.onsetSamples) return this.none();
    return this.beginUtterance();
  }

  beginUtterance() {
    this.state = StreamingEndpointState.speechActive;
    this.utteranceDurationSamples = this.configuration.preRollSamples + this.onsetEvidenceSamples;
    this.accumulatedSilenceSamples = 0;
    this.resumeEvidenceSamples = 0;
    this.ambiguousEvidenceSamples = 0;
    return {
      type: 'startUtterance',
      withPreRollSamples: this.configuration.preRollSamples,
    };
  }

  processActive(evidence, durationSamples) {
    this.utteranceDurationSamples += durationSamples;
    switch (evidence) {
      case StreamingEndpointEvidence.strong:
      case StreamingEndpointEvidence.continuing:
        this.accumulatedSilenceSamples = 0;
        return this.continue();
      case StreamingEndpointEvidence.ambiguous:
        return this.continue();
      case StreamingEndpointEvidence.silence:
        this.accumulatedSilenceSamples += durationSamples;
        if (this.accumulatedSilenceSamples < this.configuration.endEvidenceSamples) {
          return this.continue();
        }
        this.state = StreamingEndpointState.endPending;
        return this.continue();
      default:
        return this.continue();
    }
  }

  processEndPending(evidence, durationSamples) {
    this.utteranceDurationSamples += durationSamples;
    switch (evidence) {
      case StreamingEndpointEvidence.ambiguous:
        this.ambiguousEvidenceSamples += durationSamples;
        if (this.ambiguousEvidenceSamples < this.configuration.ambiguousRescueSamples) {
          return this.continue();
        }
        this.state = StreamingEndpointState.speechActive;
        this.accumulatedSilenceSamples = 0;
        this.resumeEvidenceSamples = 0;
        this.ambiguousEvidenceSamples = 0;
        return this.continue();
      case StreamingEndpointEvidence.silence:
        this.ambiguousEvidenceSamples = 0;
        this.accumulatedSilenceSamples += durationSamples;
        this.resumeEvidenceSamples = 0;
        if (this.accumulatedSilenceSamples < this.configuration.endpointSilenceSamples) {
          return this.continue();
        }
        this.reset();
        return this.finalize(StreamingEndpointFinalizeKind.endpoint);
      case StreamingEndpointEvidence.strong:
      case StreamingEndpointEvidence.continuing:
        // Resume evidence pauses the silence timer but cannot cancel the
        // pending end until the resume threshold is reached.
        this.ambiguousEvidenceSamples = 0;
        this.resumeEvidenceSamples += durationSamples;
        if (this.resumeEvidenceSamples < this.configuration.resumeSamples) {
          return this.continue();
        }
        this.state = StreamingEndpointState.speechActive;
        this.accumulatedSilenceSamples = 0;
        this.resumeEvidenceSamples = 0;
        return this.continue();
      default:
        return this.continue();
    }
  }

  none() {
    return { type: 'none' };
  }

  continue() {
    return { type: 'continueUtterance' };
  }

  finalize(kind) {
    return { type: 'finalizeUtterance', kind };
  }
}
